// @nexus/sdk/src/envelope.ts
/**
 * Nexus Envelope (Task 2.3 - Updated for v0.1.7)
 * Defines the strict NATP v0.1 data structure.
 * Handles canonical serialization and signing.
 * 
 * v0.1.7 Updates:
 * - Enhanced error handling with custom exceptions
 * - Better validation
 * 
 * NOTE: This module is kept for potential future use and backward compatibility.
 * The current transaction protocol (v0.1.7) uses a flat JSON structure instead
 * of wrapping everything in an envelope.
 */

import stringify from 'fast-json-stable-stringify';
import { v4 as uuidv4 } from 'uuid';
import sodium from 'libsodium-wrappers';
import { IdentityManager } from './identity';
import { NexusValidationError } from './exceptions';

// Define strict types for priority
export type NexusPriority = 'normal' | 'high' | 'critical';

export interface SenderInfo {
  public_key: string; // PEM format
  agent_id?: string;
}

export interface SettlementInfo {
  amount: number;
  currency: string;
  facilitation_fee: number;
}

export class NexusEnvelope {
  natp_version: string = "0.1.0";
  id: string;
  priority: NexusPriority;
  timestamp: number;
  sender: SenderInfo;
  payload: Record<string, any>;
  settlement: SettlementInfo;
  signature?: string;

  constructor(
    sender: SenderInfo,
    payload: Record<string, any>,
    priority: NexusPriority = 'normal'
  ) {
    // Validate priority
    if (!['normal', 'high', 'critical'].includes(priority)) {
      throw new NexusValidationError(
        `Invalid priority: ${priority}. Must be 'normal', 'high', or 'critical'.`
      );
    }

    this.id = uuidv4();
    this.priority = priority;
    // Python uses float seconds, JS uses ms. Convert to seconds.
    this.timestamp = Date.now() / 1000;
    this.sender = sender;
    this.payload = payload;
    this.settlement = { amount: 0, currency: 'USD', facilitation_fee: 0 };
  }

  /**
   * Returns the canonical JSON bytes of the envelope WITHOUT the signature.
   */
  public getCanonicalJson(): Uint8Array {
    const { signature, ...dataToSign } = this;
    // fast-json-stable-stringify produces compact JSON (no spaces),
    // matching Python's separators=(',', ':')
    const jsonStr = stringify(dataToSign);
    return new TextEncoder().encode(jsonStr);
  }

  /**
   * Signs the envelope using the provided IdentityManager.
   */
  public async sign(identity: IdentityManager): Promise<void> {
    const bytes = this.getCanonicalJson();
    this.signature = await identity.sign(bytes);
  }

  /**
   * Helper to parse a PEM public key back to Uint8Array for verification.
   * FIX: We must strip the ASN.1 header to get the raw Ed25519 key.
   */
  private static pemToBytes(pem: string): Uint8Array {
    // 1. Clean up the PEM string
    const b64 = pem
      .replace('-----BEGIN PUBLIC KEY-----', '')
      .replace('-----END PUBLIC KEY-----', '')
      .replace(/\s/g, '');

    // 2. Decode Base64
    const fullBytes = sodium.from_base64(b64, sodium.base64_variants.ORIGINAL);

    // 3. Strip ASN.1 Header (Fix)
    // The SubjectPublicKeyInfo format is 44 bytes total (12 bytes header + 32 bytes key).
    // Libsodium only wants the raw 32 bytes.
    if (fullBytes.length > 32) {
      return fullBytes.slice(fullBytes.length - 32);
    }

    return fullBytes;
  }

  /**
   * Verifies the envelope's signature against its own sender public key.
   */
  public async verify(): Promise<boolean> {
    if (!this.signature) {
      throw new NexusValidationError('Envelope has no signature');
    }

    await sodium.ready;

    try {
      const canonicalBytes = this.getCanonicalJson();
      const publicKeyBytes = NexusEnvelope.pemToBytes(this.sender.public_key);

      return IdentityManager.verify(canonicalBytes, this.signature, publicKeyBytes);
    } catch (e) {
      throw new NexusValidationError(`Verification failed: ${e}`);
    }
  }
}

// DX ALIAS: For simpler imports
export const Envelope = NexusEnvelope;
